hogben/utils.py

```diff
@@ -163,77 +157,182 @@ def fisher(qs: list[np.ndarray],
-
-    # Calculate the reflectance for each model for the given Q values.
-    r = np.concatenate([reflectivity(q, model)
-                        for q, model in list(zip(qs, models))])
-
-    # Calculate the Fisher information matrix using equations from the paper.
-    M = np.diag(np.concatenate(counts) / r, k=0)
-    g = np.dot(np.dot(J.T, M), J)
-
-    # If there are multiple parameters,
-    # scale each parameter's information by its "importance".
-    if len(xi) > 1:
-        if isinstance(xi[0], refnx.analysis.Parameter):
-            lb = np.array([param.bounds.lb for param in xi])
-            ub = np.array([param.bounds.ub for param in xi])
-
-        elif isinstance(xi[0], bumps.parameter.Parameter):
-            lb = np.array([param.bounds.limits[0] for param in xi])
-            ub = np.array([param.bounds.limits[1] for param in xi])
-
-        # Scale each parameter with their specified importance,
-        # scale with one if no importance was specified.
-        importance_array = []
-        for param in xi:
-            if hasattr(param, "importance"):
-                importance_array.append(param.importance)
-            else:
-                importance_array.append(1)
-        importance = np.diag(importance_array)
+
+    def __init__(self,
+                 qs: list[np.ndarray],
+                 xi: list[Union['refnx.analysis.Parameter',
+                                'bumps.parameter.Parameter']],
+                 counts: list[int],
+                 models: list[Union['refnx.reflect.ReflectModel',
+                                    'refl1d.experiment.Experiment']],
+                 step: float = 0.005):
+        """Constructor for the Fisher matrix class.
+
+        Args:
+            qs: The Q points for each model.
+            xi: The varying model parameters.
+            counts: incident neutron counts corresponding to each Q value.
+            models: models to calculate gradients with.
+            step: step size to take when calculating gradient.
+            g: The Fisher information matrix
+            min_eigenval: The minimum eigenvalue associated with the Fisher
+             information matrix
+        """
+        self.qs = qs
+        self.xi = xi
+        self.counts = counts
+        self.models = models
+        self.step = step
+
+    @property
+    def n(self) -> int:
+        """The total amount of datapoints.
+
+        Returns:
+            int: total amount of datapoints.
+        """
+        return sum(len(q) for q in self.qs)
+
+    @property
+    def m(self) -> int:
+        """The total amount of parameters.
+
+        Returns:
+            int: total amount of parameters.
+        """
+        return len(self.xi)
+
+    @property
+    def g(self) -> np.ndarray:
+        """The Fisher information matrix.
+
+        Returns:
+            numpy.ndarray: The Fisher information matrix.
+        """
+        return self._calculate_g()
```

Comment:
I'm thinking this should be 
```
self._g = self._calculate_g()
return self._g
```

Reply:
Especially now I've gotten rid of the setter, I am not really using the reference to g anywhere in the code. Also for the eigenvalue for instance, it runs the calculation again to make sure that everything is still up-to-date with the latest attributes. So a hidden self._g variable may be redundant, but perhaps it is the right practice to use it anyway.

Using `self._g` here, does allow to refer to self._g in the other methods. And for instance get rid of the argument in  `_scale_units`. This could then be done like:

```
   def __init__(self,
                 qs: list[np.ndarray],
                 xi: list[Union['refnx.analysis.Parameter',
                                'bumps.parameter.Parameter']],
                 counts: list[int],
                 models: list[Union['refnx.reflect.ReflectModel',
                                    'refl1d.experiment.Experiment']],
                 step: float = 0.005):
        """Constructor for the Fisher matrix class.

        Args:
            qs: The Q points for each model.
            xi: The varying model parameters.
            counts: incident neutron counts corresponding to each Q value.
            models: models to calculate gradients with.
            step: step size to take when calculating gradient.
            fisher_information: The Fisher information matrix
            min_eigenval: The minimum eigenvalue associated with the Fisher
             information matrix
        """
        self.qs = qs
        self.xi = xi
        self.counts = counts
        self.models = models
        self.step = step
        self._g = None

    @property
    def n(self) -> int:
        """The total amount of datapoints.

        Returns:
            int: total amount of datapoints.
        """
        return sum(len(q) for q in self.qs)

    @property
    def m(self) -> int:
        """The total amount of parameters.

        Returns:
            int: total amount of parameters.
        """
        return len(self.xi)

    @property
    def fisher_information(self) -> np.ndarray:
        """The Fisher information matrix.

        Returns:
            numpy.ndarray: The Fisher information matrix.
        """
        self._g = self._calculate_fisher_information()
        return self._g

    @property
    def min_eigenval(self) -> float:
        """The minimum eigenvalue associated with the Fisher information
         matrix.

        Returns:
            float: The minimum eigenvalue.
        """
        return np.linalg.eigvalsh(self.fisher_information)[0]

    def _calculate_fisher_information(self) -> np.ndarray:
        """Calculates the Fisher information matrix using the class attributes.

        Returns:
            numpy.ndarray: The Fisher information matrix.
        """
        if self.n == 0:
            return np.zeros((self.m, self.m))
        J = self._get_gradient_matrix()

        # Calculate the reflectance for each model for the given Q values.
        r = np.concatenate([reflectivity(q, model)
                            for q, model in list(zip(self.qs, self.models))])

        # Calculate the Fisher information matrix using equations from
        # the paper.
        M = np.diag(np.concatenate(self.counts) / r, k=0)
        self._g = np.dot(np.dot(J.T, M), J)

        # Perform unit scaling if there's at least one parameter
        if len(self.xi) >= 1:
            self._scale_units()  # Scale by unit bounds
            self._scale_importance()  # Scale by importance

    def _scale_units(self):
        """Scale the Fisher information matrix using unit scaling.

        Args:
            g: The Fisher information matrix.

        Returns:
            numpy.ndarray: The scaled Fisher information matrix.
        """
        lb, ub = self._get_bounds()
        H = np.diag(1 / (ub - lb))  # Get unit scaling Jacobian.
        self._g = np.dot(np.dot(H.T, g), H)  # Perform unit scaling.

    def _scale_importance(self):
        """Scale the Fisher information matrix using importance scaling.

        Args:
            g: The Fisher information matrix.

        Returns:
            numpy.ndarray: The scaled Fisher information matrix.
        """
        importance_array = [param.importance if hasattr(param, "importance")
                            else 1 for param in self.xi]
        importance = np.diag(importance_array)
        self._g = np.dot(self._g, importance)
```

Not 100% what is considered cleaner, but both work.

Reply:
For now I decided to not use the hidden `self._g` variable, I find the `calculate_fisher` method becomes a bit more readable, with a more explicit flow of data around the scaling part. 

But I did get rid of the redundant setter. (Actually, the whole code would crash if I were to set the FI matrix directly due to a recursion error).